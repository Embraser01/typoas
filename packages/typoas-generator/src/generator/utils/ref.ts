import { Expression, factory, TypeNode } from 'typescript';

export const IMPORT_RUNTIME = 'r';

export enum ExportedRef {
  HttpMethod = 'HttpMethod',
  BaseServerConfiguration = 'BaseServerConfiguration',
  HttpLibrary = 'HttpLibrary',
  RefResolver = 'RefResolver',
  IsomorphicFetchHttpLibrary = 'IsomorphicFetchHttpLibrary',
  SchemaRefResolver = 'SchemaRefResolver',
  ApiException = 'ApiException',
  AuthProvider = 'AuthProvider',
  BasicAuthConfig = 'BasicAuthConfig',
  BearerAuthConfig = 'BearerAuthConfig',
  BaseFlowConfig = 'BaseFlowConfig',
  SecurityAuthentication = 'SecurityAuthentication',
  ApiKeySecurityAuthentication = 'ApiKeySecurityAuthentication',
  HttpSecurityAuthentication = 'HttpSecurityAuthentication',
  OAuth2SecurityAuthentication = 'OAuth2SecurityAuthentication',
  applyTemplating = 'applyTemplating',
  applyTransforms = 'applyTransforms',
  isCodeInRange = 'isCodeInRange',
  serializeParameter = 'serializeParameter',
}

export function createRuntimeRefProperty(exportedRef: ExportedRef): Expression {
  return factory.createPropertyAccessExpression(
    factory.createIdentifier(IMPORT_RUNTIME),
    exportedRef,
  );
}

export function createRuntimeRefType(
  exportedRef: ExportedRef,
  typeArguments?: readonly TypeNode[],
): TypeNode {
  return factory.createTypeReferenceNode(
    factory.createQualifiedName(
      factory.createIdentifier(IMPORT_RUNTIME),
      factory.createIdentifier(exportedRef),
    ),
    typeArguments,
  );
}

export function createJSONParseWrapper(data: unknown): Expression {
  return factory.createCallExpression(
    factory.createPropertyAccessExpression(
      factory.createIdentifier('JSON'),
      'parse',
    ),
    undefined,
    [factory.createStringLiteral(JSON.stringify(data), true)],
  );
}
