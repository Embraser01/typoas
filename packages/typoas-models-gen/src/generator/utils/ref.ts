import { Expression, factory, TypeNode } from 'typescript';

export const IMPORT_RUNTIME = 'runtime';

export enum ExportedRef {
  HttpMethod = 'HttpMethod',
  BaseServerConfiguration = 'BaseServerConfiguration',
  HttpLibrary = 'HttpLibrary',
  SchemaRefResolver = 'SchemaRefResolver',
  ApiException = 'ApiException',
  applyTemplating = 'applyTemplating',
  applyTransforms = 'applyTransforms',
  isCodeInRange = 'isCodeInRange',
}

export function createRuntimeRefProperty(exportedRef: ExportedRef): Expression {
  return factory.createPropertyAccessExpression(
    factory.createIdentifier(IMPORT_RUNTIME),
    exportedRef,
  );
}

export function createRuntimeRefType(exportedRef: ExportedRef): TypeNode {
  return factory.createTypeReferenceNode(
    factory.createQualifiedName(
      factory.createIdentifier(IMPORT_RUNTIME),
      factory.createIdentifier(exportedRef),
    ),
  );
}
