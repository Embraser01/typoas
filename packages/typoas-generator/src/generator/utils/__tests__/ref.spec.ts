import { describe, expect, it } from '@jest/globals';
import { SchemaObject } from 'openapi3-ts';
import { filterSchema } from '../ref';

describe('filter schema', () => {
  it('should handle allOf', () => {
    const schema: SchemaObject = {
      allOf: [
        {
          type: 'object',
          properties: { d: { type: 'string', title: 'ObjectD' } },
        },
        {
          type: 'object',
          properties: { a: { type: 'string', title: 'ObjectA' } },
        },
      ],
    };

    expect(filterSchema(schema)).toMatchSnapshot();
  });

  it('should handle additionalProperties', () => {
    const schema: SchemaObject = {
      type: 'object',
      additionalProperties: {
        type: 'object',
        properties: { a: { type: 'string', title: 'ObjectA' } },
      },
    };

    expect(filterSchema(schema)).toMatchSnapshot();
  });

  it('should handle additionalProperties true', () => {
    const schema: SchemaObject = {
      type: 'object',
      additionalProperties: true,
    };

    expect(filterSchema(schema)).toMatchSnapshot();
  });
});
