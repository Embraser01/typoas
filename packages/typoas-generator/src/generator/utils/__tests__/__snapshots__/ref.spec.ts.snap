// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter schema should handle additionalProperties 1`] = `
Object {
  "additionalProperties": Object {
    "properties": Object {
      "a": Object {
        "type": "string",
      },
    },
    "type": "object",
  },
  "type": "object",
}
`;

exports[`filter schema should handle additionalProperties true 1`] = `
Object {
  "additionalProperties": true,
  "type": "object",
}
`;

exports[`filter schema should handle allOf 1`] = `
Object {
  "allOf": Array [
    Object {
      "properties": Object {
        "d": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
    Object {
      "properties": Object {
        "a": Object {
          "type": "string",
        },
      },
      "type": "object",
    },
  ],
}
`;
